### PR TITLE
Async validation support

### PR DESCRIPTION
**Pull request by elhigu**
This should be released as major release just to be safe.

### CODE REVIEW

File: src/Pool.ts [hunk 1]
```diff
@@ -2,6 +2,8 @@ import { PendingOperation } from './PendingOperation';
 import { Resource } from './Resource';
 import { checkOptionalTime, delay, duration, now, reflect, tryPromise } from './utils';
 import { EventEmitter } from 'events';
+import { AssertionError } from 'assert';
```
Comment (kibertoad) [inline src/Pool.ts]: What is this for?

File: tests.js [hunk 2]
```diff
@@ -704,6 +690,297 @@ describe('Tarn', () => {
...
+      },
+      destroy(resource) {
+        ++destroyCalled;
+      },
+      min: 0,
+      max: 2,
+      acquireTimeoutMillis: 100
+    });
+
+    return pool
```
Comment (kibertoad) [inline tests.js]: Why not async/await? We are at Node 8 already for Tarn

File: src/Pool.ts [hunk 3]
```diff
@@ -339,32 +382,80 @@ export class Pool<T> {
...
+                this._tryAcquireOrCreate();
+              }, 0);
+            }
+          } finally {
+            remove(this.pendingValidations, pendingAcquire);
+          }
+        });
+
+      // try to validate
+      this._validateResource(free.resource)
```
Comment (kibertoad) [inline src/Pool.ts]: we can use async/await

File: src/Pool.ts [hunk 4]
```diff
@@ -374,13 +465,10 @@ export class Pool<T> {
 
   _validateResource(resource: T) {
     try {
-      return !!this.validate(resource);
+      return Promise.resolve(this.validate(resource));
```
Comment (kibertoad) [inline src/Pool.ts]: you can make function async, and then it will wrap results of awaited function into promise automatically for you.

File: src/Pool.ts [hunk 5]
```diff
@@ -339,32 +382,80 @@ export class Pool<T> {
...
+          try {
+            if (validationSuccess) {
+              // At least one active resource exist, start reaping.
+              this._startReaping();
+              pendingAcquire.resolve(free.resource);
+            } else {
+              remove(this.used, free);
+              this._destroy(free.resource);
+
+              // is aquire was canceled, failed or timed out already
```
Comment (kibertoad) [inline src/Pool.ts]: aquire -> acquire

File: src/Pool.ts (hunk 1 above)
Comment (elhigu) [reply src/Pool.ts]: Looks like that is reminder of early phase of implementation, when there was some code to narrow typing. Will remove it.

File: src/Pool.ts (hunk 3 above)
Comment (elhigu) [reply src/Pool.ts]: I didn't want to go there yet. I'd rather have that change done as a separate PR and change it completely. 

Actually this code is already written in typescript so we could have done that also earlier and compile it for node 6. Though I'm glad that we didn't since I actually had to do lots of running the code in debugger to figure out why stuff didn't work and to figure out how it should work. Code didn't seem to have source maps set properly, so this way it was lot easier to follow up the execution from compiled code.

File: src/Pool.ts (hunk 4 above)
Comment (elhigu) [reply src/Pool.ts]: I didn't want to mix async / await in yet in this PR.

File: tests.js (hunk 2 above)
Comment (elhigu) [reply tests.js]: I didn't want to mix async / await in yet in this PR.

File: src/Pool.ts (hunk 5 above)
Comment (elhigu) [reply src/Pool.ts]: Will fix

Comment (elhigu) [reply src/Pool.ts]: done

**Comment by elhigu** [comment]
Thanks for review @kibertoad 👍  I'll still wait for a few days before merging if someone wants to figure out if there are some more corner cases that I should test or try to spot errors in the new doAcquire logic.
